File: unijava/src/org/unified/uni/CharReader.java
package org.unified.uni;

/**
 * Text reader
 * <p>
 * Created by Michael Vlasov on 19.04.2018.
 * <p>
 * Copyright (c) 2017 Michael Vlasov
 */
public class CharReader {
	public interface Test {
		boolean apply(char c);
	}


	public CharReader(String source) {
		this.source = source;
		next_pos = 0;
		current_pos = 0;
		length = source.length();
		has_current = true;
		tag = null;
		current = '\0';
		passed_start = 0;
		passed_end = 0;
		next();
	}


	public void next() {
		if (!has_current) {
			return;
		}
		if (next_pos < length) {
			current_pos = next_pos;
			current = source.charAt(next_pos);
			next_pos += 1;
		} else {
			has_current = false;
			current = '\0';
			current_pos = length;
		}
	}


	public String context_at(int pos) {
		Context context = new Context(pos);
		return "\r\n" +
			"line: " + (context.line + 1) + ", col: " + (context.index - context.line_start + 1) + "\r\n" +
			"here: " +
			source.substring(context.line_start, context.index) +
			"👉" + source.substring(context.index, context.line_end);
	}


	public String current_context() {
		return context_at(current_pos);
	}


	public boolean pass_char(char test) {
		if (has_current && current == test) {
			next();
			return true;
		}
		return false;
	}


	public boolean pass_while(Test test) {
		int start_pos = current_pos;
		while (has_current && test.apply(current)) {
			next();
		}
		int end_pos = current_pos;
		if (end_pos == start_pos) {
			return false;
		}
		passed_start = start_pos;
		passed_end = end_pos;
		return true;
	}


	public String passed() {
		return source.substring(passed_start, passed_end);
	}

	// Internals

	private String source;
	int next_pos;
	protected int current_pos;
	private int length;
	boolean has_current;
	Object tag;
	char current;
	private int passed_start;
	private int passed_end;

	private class Context {
		int line;
		int line_start;
		int line_end;
		int index;


		Context(int index) {
			line = 0;
			line_start = 0;
			int i = 0;
			while (i < length) {
				char c = source.charAt(i);
				if (c == '\r' || c == '\n') {
					line_end = i;
					i += 1;
					if (i < length && c == '\r' || source.charAt(i) == '\n') {
						i += 1;
					}
					if (i >= index) {
						this.index = Math.min(index, line_end);
						return;
					}
					line_start = i;
					line += 1;
				} else {
					i += 1;
				}
			}
			line_end = length;
			this.index = Math.min(index, length);
		}

	}
}

File: uniswift/Uni/Uni/Lexer/CharReader.swift
//
//  CharReader.swift
//  Uni
//
//  Created by Michael Vlasov on 26.04.2018.
//  Copyright © 2018 Unified. All rights reserved.
//

import Foundation

public final class CharReader {
	public let string: String
	public let unicode: String.UnicodeScalarView

	public private(set) var hasCurrent: Bool
	public private(set) var current: UnicodeScalar
	public private(set) var currentPos: String.UnicodeScalarView.Index
	public private(set) var nextPos: String.UnicodeScalarView.Index
	public var tag: Any?

	public init(string: String) {
		self.string = string
		unicode = string.unicodeScalars
		nextPos = unicode.startIndex
		currentPos = nextPos
		endPos = unicode.endIndex
		hasCurrent = true
		current = "\0"
		passedStart = currentPos
		passedEnd = currentPos
		next()
	}
	
	
	public func next() {
		if !hasCurrent {
			return
		}
		if nextPos < endPos {
			currentPos = nextPos
			current = unicode[nextPos]
			nextPos = unicode.index(after: nextPos)
		} else {
			hasCurrent = false
			current = "\0"
			currentPos = endPos
		}
	}


	public func contextAt(pos: String.UnicodeScalarIndex) -> String {
		let context = Context(source: unicode, index: pos)
		return "\r\n" +
			"line: \(context.line + 1), col: \(unicode.distance(from: context.index, to: context.lineStart) + 1)\r\n" +
			"here: \(string[context.lineStart ..< context.index])👉\(string[context.index ..< context.lineEnd])"
	}

	
	@discardableResult
	public func passChar(_ test: UnicodeScalar) -> Bool {
		if hasCurrent && current == test {
			next()
			return true
		}
		return false
	}

	
	@discardableResult
	public func passWhile(_ test: (UnicodeScalar) -> Bool) -> Bool {
		let startPos = currentPos
		while hasCurrent && test(current) {
			next()
		}
		let endPos = currentPos
		if endPos == startPos {
			return false
		}
		passedStart = startPos
		passedEnd = endPos
		return true
	}
	
	
	public var passed: Substring {
		return string[passedStart ..< passedEnd]
	}

	
	public var currentContext: String  {
		return contextAt(pos: currentPos)
	}

	// MARK: - Internals
	
	
	private var endPos: String.UnicodeScalarView.Index
	private var passedStart: String.UnicodeScalarView.Index
	private var passedEnd: String.UnicodeScalarView.Index

	private struct Context {
		let line: Int
		let lineStart: String.UnicodeScalarView.Index
		let lineEnd: String.UnicodeScalarView.Index
		let index: String.UnicodeScalarView.Index


		init(source: String.UnicodeScalarView, index: String.UnicodeScalarView.Index) {
			var line = 0
			var i = source.startIndex
			let end = source.endIndex
			var lineStart = i
			while i < end {
				let c = source[i]
				if c == "\r" || c == "\n" {
					let lineEnd = i
					i = source.index(after: i)
					if i < end && c == "\r" || source[i] == "\n" {
						i = source.index(after: i)
					}
					if i >= index {
						self.line = line
						self.lineStart = lineStart
						self.lineEnd = lineEnd
						self.index = min(index, lineEnd)
						return
					}
					lineStart = i
					line += 1
				} else {
					i = source.index(after: i)
				}
			}
			self.line = line
			self.lineStart = lineStart
			self.lineEnd = end
			self.index = min(index, end)
		}

	}

}
